### PR TITLE
Link to the Steam store rather than to ModBot

### PR DESCRIPTION
**Pull request by woctezuma**
http://www.neogaf.com/forum/showpost.php?p=230877232&postcount=8715

### CODE REVIEW

File: neogaf-monkeybot.user.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ var lastUpdate = localStorage.getItem("monkeyBot_steamGameListUpdatedOn") || "";
 var modBotUrl = "http://www.neogaf.com/forum/private.php?do=newpm&u=253996";
 var modBotPosts = $("[data-username='ModBot']");
 var modBotSelfPosts = $("a[href='member.php?u=253996']").closest(".postbit").find(".post");
-var allPosts = modBotPosts.add(modBotSelfPosts);
+var allPosts = modBotPosts.add(modBotSelfPosts);var storeUrl = "http://store.steampowered.com/search/?term=";
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Newline between `;` and `var`

**Comment by woctezuma** [comment]
I could make the ModBot code clickable. Unfortunately, I have no idea how to auto-fill the PM with the ModBot code, 🤔

Regarding the redirection to the store page, I don't know what the simplest way to do it would be. Redirecting to the search page is easy, but getting the appId based on the appName requires some knowledge which I don't have... unless I make a request to steamDB or something. 😅

**Comment by petetnt** [comment]
Actually the `appId`s are already fetched when the call is made to `http://api.steampowered.com/IPlayerService/GetOwnedGames/v0001/`

The response looks like this:

```json
{
	"response": {
		"game_count": 1369,
		"games": [
			{
				"appid": 10,
				"name": "Counter-Strike",
				"playtime_forever": 153,
				"img_icon_url": "6b0312cda02f5f777efa2f3318c307ff9acafbb5",
				"img_logo_url": "af890f848dd606ac2fd4415de3c3f5e7a66fcb9f",
				"has_community_visible_stats": true
			},
			{
				"appid": 80,
				"name": "Counter-Strike: Condition Zero",
				"playtime_forever": 0,
				"img_icon_url": "077b050ef3e89cd84e2c5a575d78d53b54058236",
				"img_logo_url": "acdb28ba1b4c2fcc526332c1b63fc0f7533f087f"
			},
			{
				"appid": 100,
				"name": "Counter-Strike: Condition Zero Deleted Scenes",
				"playtime_forever": 0,
				"img_icon_url": "077b050ef3e89cd84e2c5a575d78d53b54058236",
				"img_logo_url": "acdb28ba1b4c2fcc526332c1b63fc0f7533f087f"
			},
                { ... }
         ]
}
```

We filter the list to just names in https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L102 but instead of returning just the names in the map call we could return a object of

``` js
{
  "appId",
  "sanitizedName" 
}
```

and use those in the comparison part.

**Comment by petetnt** [comment]
For the ModBot autofill, it works like this:

1. When you click the link, the script saves the code to `localStorage` as `raffleLine`
2. When you navigate to private messages, the checks if there's `raffleLines` and fills them into the fields:
https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L227

I remember that there's some bug in `vbulletin` that prevents the filling of the title field, but those can be handled in separate issue.

**Comment by woctezuma** [comment]
Alright, I made the ModBot code clickable and the ModBot autofill is working for me for some reason. However, I don't know how to remove the end of the line in the original post, so the ModBot code is repeated twice, one clickable and once not clickable.

**Comment by woctezuma** [comment]
For the map of game and appId, I get the idea but I don't know javascript so I'm really stuck.

**Comment by petetnt** [comment]
Basically you need to modify this line:

https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L102
So it returns the name of the game and the appid:

``` js
    ownedGames = games.map(function map(game) {
        return {
           appid: game.appid,
           sanitizedName: sanitizeName(game.name),
       }
    });
```

Then you need to check where the `ownedGames` variable is used, for example in 

https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L40

This needs to be modified so that instead of searching for the `indexOf` the name in the array we search for the name in the array of objects.

For example something like

``` js
function checkIfOwnedOnSteam(name, line) {
    return ownedGames.find(game => {
      return game.name === name;
    })  && !/uPlay|\(GoG\)|\(Origin\)|Desura/.test(line);
}
```

Now that because we need the `appid`, we can use the result of the `checkIfOwnedOnSteam` function here:

https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L68

Lets modify the `checkIfOwnedOnSteam` a bit more:

``` js
function getIfOwnedOnSteam(name, line) {
    if (!/uPlay|\(GoG\)|\(Origin\)|Desura/.test(line)) {
      return false;
    }

    return ownedGames.find(game => {
      return game.name === name;
    })
}
```

Now if the 

- Game is not a Steam game, it returns `false`
- If the game is owned, it returns an object of `{ appid, sanitizedName }`
- If the game is not owned, it returns `null`

so we can use it on Line 68 like this:

``` js
var game = getIfOwnedOnSteam(name, line) 

if (game) {
  /** inside this block we can access the appid of the game with game.appid **/
  
  <div>game.appid</div>
}
```

**Comment by woctezuma** [comment]
Regarding modbot code autofill, I have figured out why it works for me now: I use left-click. Indeed, I used to middle-clicke the links, in which case it does not work.

Regarding the appid links, I get that it can only work for owned games. I will try to make it work using your help, but I think it is less useful than the link for games not yet owned: typically, one wants to check the store pages of games not yet owned to choose which giveaway to enter.

**Comment by petetnt** [comment]
@woctezuma Haha, oh wow, sorry I am an idiot (or it was an early morning before coffee). I can see the difficulty of getting the ID from the `ownedGames` when you don't own the game 😹 😂 

You could retrieve the whole game list from http://api.steampowered.com/ISteamApps/GetAppList/v0001/ and filter & cache those the same way the `ownedGames` list is.

**Comment by woctezuma** [comment]
Argh, not working! I don't know how to debug javascript, and it is failing at some point.

Cannot I parse JSON with a line such as:
`parseAllGames(JSON.parse(response.responseText).applist.apps.app);`

**Comment by petetnt** [comment]
I think 
```
JSON.parse(response.responseText).response.applist.apps.app
```

would give you the app list in `json`.

see: https://github.com/petetnt/neogaf-monkeybot/blob/master/neogaf-monkeybot.user.js#L197

For debugging `console.log` is a good start. Open your developer tools with F12 (or cmd-opt-j on OSX) and see the console.

Add some lines like `console.log(JSON.parse(response.responseText))` and you should see the output in the console when you run the script. Debugging Greasemonkey scripts are bit a pain though.

**Comment by woctezuma** [comment]
I thought the response keyword was specific to the owned games.

For the owned game, the page looks like:

`{
	"response": {
		"game_count": 1719,
		"games": [
			{
				"appid": 1200,
				"name": "Red Orchestra: Ostfront 41-45",
				"playtime_forever": 0,
				"img_icon_url": "66519d69efc5b98bdafe61347db4ec8be01dff7b",
				"img_logo_url": "7c41b2bfa0952f44959e77b2650d09a7a22e633a",
				"has_community_visible_stats": true
			},`

For the whole list of Steam games, the page looks like:

`{
	"applist": {
		"apps": {
			"app": [
				{
					"appid": 5,
					"name": "Dedicated Server"`

**Comment by petetnt** [comment]
Hmm, you are correct. What kinda error is your first line throwing? You should see it in the console too

**Comment by woctezuma** [comment]
I don't know, I have a weird message:

`The Web Console logging API (console.log, console.info, console.warn, console.error) has been disabled by a script on this page.`

Anyway, I think the problem is at the initialization. With your original script, the script asks for the API key. With mine, it does not. Plus local storage does not contain the new variables allGames, etc.

**Comment by petetnt** [comment]
I can try and check it out tomorrow 👍

**Comment by woctezuma** [comment]
I have tested on a minimal example and it the JSON parsing works in both cases. I must have made a stupid mistake somewhere else.

Interestingly, Firefox had trouble parsing the JSON for the whole Steam store with [the link you provided](http://api.steampowered.com/ISteamApps/GetAppList/v0001/), and I had to Alt+F4 it.

**Comment by woctezuma** [comment]
It looks like I did not commit to the right branch. I am doing it now.

**Comment by woctezuma** [comment]
The Modbot link and autofill works fine.

However, the code as is should be able to link to the Steam store and fails. It seems the reason why is that the appid is "undefined". So the current implementation defaults to using Steam search with the name of the game. If you could pinpoint where the use of the object pair fails, it would solve the problem.

**Comment by woctezuma** [comment]
Actually, it seems the problem arises from the result of getIfOnSteam(). When I try it by hand, it works. When the script runs, it does not return the right object.

**Comment by woctezuma** [comment]
Problem fixed. There was a wrong `!` in the function `getIfOwnedOnSteam `you had given to me! Silly boy! 😝

Please accept the pull request! 👌

**Comment by woctezuma** [comment]
I have also fixed the treatment of games with special characters such as "Cosmic Dust & Rust" on the first line of this giveaway: http://www.neogaf.com/forum/showpost.php?p=231131849&postcount=10309

File: neogaf-monkeybot.user.js [hunk 2]
```diff
@@ -53,6 +75,23 @@ function clickHandler(event) {
     window.location.href = modBotUrl;
 }
 
+/*
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Some JSDoc formatting: 

``` js
/**
  * Escape special characters to be able to treat games such as "Cosmic Dust & Rust" with a "&"
  * See: https://stackoverflow.com/questions/784586/ and http://stackoverflow.com/questions/1787322/
  * 
  * @param {string} text - text to escape
  * @returns {string} - modified text
  */
```

File: neogaf-monkeybot.user.js [hunk 3]
```diff
@@ -53,6 +75,23 @@ function clickHandler(event) {
...
+ * See: https://stackoverflow.com/questions/784586/
+ *      http://stackoverflow.com/questions/1787322/
+ */
+function escapeHtml(text) {
+  var map = {
+    '&': '&amp;',
+    '<': '&lt;',
+    '>': '&gt;',
+    '"': '&quot;',
+    "'": '&#039;'
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Missing trailing comma

File: neogaf-monkeybot.user.js [hunk 4]
```diff
@@ -65,22 +104,44 @@ function matchGames() {
         Object.keys(giveaways || {}).forEach(function mapGiveaways(key) {
             var line = giveaways[key];
             var name = line.split("--")[0].trim();
-
+            var modbotcode = line.split("--")[1].trim();
+
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Some indentation issues from line 107 to 118:

- Spaces, not tabs
- No extra newlines
- Follow the indentation

File: neogaf-monkeybot.user.js [hunk 5]
```diff
@@ -65,22 +104,44 @@ function matchGames() {
...
+	
+		var urlToShow = storeUrl + name;
+			
+		var game = getIfOnSteam(name, line);
+		if(game) {
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Space between `if (game)`

File: neogaf-monkeybot.user.js [hunk 6]
```diff
@@ -65,22 +104,44 @@ function matchGames() {
...
                         "<span class='inLibraryFlag'>IN LIBRARY &nbsp;&nbsp</span>" +
-                        "<span class='inLibraryText'>" + name + "</span>"
+                        "<span class='inLibraryText'>" +
+				"<a class='visitSteamStorePageOwnedGame' " +
+                            	"title='Click me to visit the Steam store page of your game' " +
+                            	"href='" + urlToShow + "/'>" + name + "</a>"     
+			+ "</span>"
                     ));
             } else {
+                    $elem.html(
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Mixed indentation from L131 to L137

File: neogaf-monkeybot.user.js [hunk 7]
```diff
@@ -109,6 +169,26 @@ function parseOwnedGames(json) {
     matchGames();
 }
 
+/*
+ * Parses the whole list of Steam games from json response
+ * @param {object} json - Json response
+ */
+function parseAllGames(json) {
+
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Extra newline here

File: neogaf-monkeybot.user.js [hunk 8]
```diff
@@ -43,6 +47,24 @@ function checkIfOwnedOnSteam(name, line) {
...
+ * @param   {string}  name Name of the game
+ * @param   {line}    line Modbot line of the game
+ * @returns {map} 	  a map { appid, sanitizedName } if the game exists on the Steam store, false if not
+ */
+function getIfOnSteam(name, line) {
+    if (/uPlay|\(GoG\)|\(Origin\)|Desura/.test(line)) {
+      return false;
+    }
+
+    return allGames.find(game => {
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: This arrow function should probably be a regular named function (eventually everything here will be ES2015-8 but lets stick with `ES5` for now:

``` js
return allGames.find(game, function findGame() {
  return game.sanitizedName === sanitizeName(name);
});
````

File: neogaf-monkeybot.user.js [hunk 9]
```diff
@@ -43,6 +47,24 @@ function checkIfOwnedOnSteam(name, line) {
...
+      return game.sanitizedName === sanitizeName(name);
+    })
+}
+
+/**
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Empty comment here?

File: neogaf-monkeybot.user.js [hunk 10]
```diff
@@ -53,6 +75,23 @@ function clickHandler(event) {
...
+function escapeHtml(text) {
+  var map = {
+    '&': '&amp;',
+    '<': '&lt;',
+    '>': '&gt;',
+    '"': '&quot;',
+    "'": '&#039;'
+  };
+
+  return text.replace(/[&<>"']/g, function(m) { return map[m]; });
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: The function should be a named function or the linter will complain about it:

`function renameChar(m) { return map[m]; });`

File: neogaf-monkeybot.user.js [hunk 11]
```diff
@@ -43,6 +47,24 @@ function checkIfOwnedOnSteam(name, line) {
...
+ * @returns {map} 	  a map { appid, sanitizedName } if the game exists on the Steam store, false if not
+ */
+function getIfOnSteam(name, line) {
+    if (/uPlay|\(GoG\)|\(Origin\)|Desura/.test(line)) {
+      return false;
+    }
+
+    return allGames.find(game => {
+      return game.sanitizedName === sanitizeName(name);
+    })
```
Comment (petetnt) [inline neogaf-monkeybot.user.js]: Missing trailing `;`

**Comment by woctezuma** [comment]
Is there a way to accept your changes or do I modify the code by hand?

**Comment by petetnt** [comment]
You can just modify the code by hand and commit them, the comments that are fixed are marked as done.

File: neogaf-monkeybot.user.js (hunk 3 above)
Comment (woctezuma) [inline neogaf-monkeybot.user.js]: Not sure because it the last one of the map.

Comment (petetnt) [inline neogaf-monkeybot.user.js]: Yeah the linter prefers trailing comma. The trailing commas aren't actually an issue in any modern-ish browsers (that is > IE8) and it creates neater git changelogs and makes life easier.

There's even a proposal that will land in ES2017 (or ES2018?) that enables trailing commas for function parameters, sweet!

**Comment by woctezuma** [comment]
I have done them all, I think. There might be some indentation (spaces vs tabs) issues. Is there a way to fix these automatically?

**Comment by woctezuma** [comment]
I have broken something.

**Comment by woctezuma** [comment]
There is a problem with this commit:
https://github.com/petetnt/neogaf-monkeybot/pull/6/commits/3ec80ff2772159605af6acc75e7153068e2946ca

**Comment by petetnt** [comment]
@woctezuma whoops sorry, the correct format is

``` js
return allGames.find(function findGame(game) {
```

**Comment by woctezuma** [comment]
Alright. All done and working. I have tested here:
http://www.neogaf.com/forum/showpost.php?p=231131849&postcount=10309
and there:
http://www.neogaf.com/forum/showpost.php?p=231158649&postcount=10438

**Comment by petetnt** [comment]
LGTM, thanks for this! I'll fix couple of small things later, bump the version and publish it 👍

**Comment by woctezuma** [comment]
Thanks, so I don't need to do anything? Right? You have the rights?

**Pull request merged**